rust/src/database/verkle/crypto/commitment.rs

```diff
@@ -0,0 +1,241 @@
+use verkle_trie::constants::CRS;
+use zerocopy::{FromBytes, Immutable, IntoBytes, Unaligned};
+
+use crate::database::verkle::crypto::Scalar;
+
+/// A vector commitment to a sequence of 256 scalar values, using the Pedersen commitment scheme
+/// on the Banderwagon curve.
+///
+/// Commitments can be de-/serialized from/to bytes using zerocopy.
+/// Note that not all byte arrays correspond to valid commitments, i.e., points in the Banderwagon
+/// prime subgroup. Performing any operations on invalid commitments will produce indeterminate
+/// results (garbage in, garbage out).
+#[derive(Debug, Clone, PartialEq, Eq, FromBytes, IntoBytes, Immutable, Unaligned)]
+#[repr(C)]
+pub struct Commitment {
+    // We store a byte representation to allow for easy serialization using zerocopy.
+    // Note that conversion from/to banderwagon::Element is NOT trivial (due to compression).
+    // TODO: Benchmark performance impact (https://github.com/0xsoniclabs/sonic-admin/issues/373).
+    point_bytes: [u8; 32],
+}
+
+// Creating the committer is very expensive (in the order of seconds!), so we cache it.
+static COMMITTER: LazyLock<DefaultCommitter> = LazyLock::new(|| DefaultCommitter::new(&CRS.G));
+
+impl Commitment {
+    /// Creates a commitment to the given sequence of scalar values.
+    /// At most 256 values are used, any additional values are ignored.
+    /// If fewer than 256 values are provided, the remaining values are equal to
+    /// the return value of [`Scalar::zero`].
+    pub fn new(values: &[Scalar]) -> Self {
+        let point = COMMITTER.commit_lagrange(&values.iter().map(Fr::from).collect::<Vec<Fr>>());
+        Commitment {
+            point_bytes: point.to_bytes(),
+        }
+    }
+
+    /// Updates the commitment by changing the value at the given index.
+    pub fn update(&mut self, index: u8, old_value: &Scalar, new_value: &Scalar) {
+        let delta = Fr::from(new_value) - Fr::from(old_value);
+        let delta_commitment = COMMITTER.scalar_mul(delta, index as usize);
+        self.point_bytes = (self.as_element() + delta_commitment).to_bytes();
+    }
+
+    /// Maps the commitment point to the Banderwagon scalar field,
+    /// allowing it to be used as input to other commitments.
+    pub fn to_scalar(&self) -> Scalar {
+        Scalar::from(self.as_element().map_to_scalar_field())
+    }
+
+    /// Returns a hash corresponding to the commitment.
+    /// Used for computing keys during Verkle trie state embedding.
+    pub fn hash(&self) -> [u8; 32] {
+        let scalar = self.as_element().map_to_scalar_field();
+        scalar
+            .into_bigint()
+            .to_bytes_le()
+            .try_into()
+            .expect("scalar field element should be 32 bytes")
+    }
+
+    /// Returns a compressed 32-byte representation of the commitment.
+    /// Used as the state root commitment in Verkle tries.
+    pub fn compress(&self) -> [u8; 32] {
+        self.point_bytes
+    }
```

Comment:
@HerbertJordan: Hash and compress produce different values and are used differently, but do you know why we could not just use one of these everywhere?